### PR TITLE
fix(team-approvals): authorize file preview reads

### PR DESCRIPTION
**Pull request by 777genius**
## Summary
- bind every approval preview read to the exact active team, run, request, and file path
- reject spoofed, stale, relative, or unrelated paths before filesystem access
- honor bytesRead on short reads to avoid zero-filled tail corruption
- move the renderer approval API into the feature contract and tighten the legacy size ratchet from 1163 to 1151 lines

## Verification
- pnpm typecheck
- focused fast lint for all changed TypeScript files
- pnpm guard:source-file-size
- 72 focused Vitest tests

## Coordination
SERIAL FREEZE: Draft only. Do not merge until the canonical merge resolver owner releases the freeze. This PR does not push or merge refactor/hosted-web-feature-boundaries or dev.

## Summary by CodeRabbit

* **New Features**
  * Added secure tool-approval “file preview” support using a structured request (team/run/request/filePath) and a pending approval file-path lookup.
  * Wired preview-reading end-to-end across IPC, preload, and renderer, and updated diff preview components to include team/run identity.
* **Bug Fixes**
  * Strengthened authorization so previews are rejected for stale, mismatched, or unauthorized identities.
  * Improved preview file reading: safe truncation from actual bytes read, correct decoding, and rejection of symlink/race-condition paths.
* **Tests**
  * Added/updated coverage for preview lookup, authorization behavior, and file-reading edge cases (including short reads and truncation).

### CODE REVIEW

**Comment by 777genius** [comment]
Hosted exact-head review evidence for `0f7b4cf2c57a9febf8fab0a477b4e088d7c7e2a1` against base `082bf7e8fd426578905efaab97645bd6ef98b31c`.

Verdict: **NO P0-P3**.

Critical findings fixed during this review cycle:
- exact approval identity now authorizes and reads the same approved path, closing lexical/symlink path substitution
- native pending approvals are bound to the tracked run
- renderer clears stale diff state before paint and refreshes collapsed quick stats for reused request IDs
- truncation uses the final open-file descriptor size and actual bytes read
- IPC depends on one application use case through narrow ports

Evidence:
- hosted focused tests: 335 passed
- typecheck: passed
- cumulative changed-file lint: passed
- source-size guard: 2996 production files, 170 legacy entries, 800-line limit
- GitHub test shards, lint splits, validate, Windows smoke, CodeQL and Analyze: all passed

The branch remains draft and unmerged because the serial canonical/dev freeze is still active.

**Comment by 777genius** [comment]
@codex review

Please review exact head `00a2b170427cefff13d927e566fbd254118dd544` against base `082bf7e8fd426578905efaab97645bd6ef98b31c`. Focus on approval identity authorization, stale async renderer state, file read boundaries, symlink/path substitution, truncation semantics, and IPC contract regressions. Report concrete P0-P3 findings only. Previous exact-head evidence is superseded by this head.

File: src/renderer/components/team/ToolApprovalDiffPreview.test.tsx [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    loading: false,
+    error: null,
+    fileName: '',
+    oldString: '',
+    newString: '',
+    isNewFile: false,
+    truncated: false,
+    isBinary: false,
+  }),
+}));
```
[the remark this answers is not in the document]
Comment (777genius) [reply src/renderer/components/team/ToolApprovalDiffPreview.test.tsx]: Fixed in `00a2b170427cefff13d927e566fbd254118dd544`. The mock now captures the hook call and asserts the exact team, run, reused request ID, file path, payload, and collapsed state. Focused approval tests pass 45/45, typecheck and changed-file lint pass.

File: src/features/team-approvals/core/application/use-cases/ReadToolApprovalFilePreview.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    const approvedPath = this.dependencies.pendingApprovals.getFilePath(
+      request.teamName,
+      request.runId,
+      request.requestId
+    );
+    if (approvedPath === null || request.filePath !== approvedPath) {
+      return null;
+    }
+
+    return this.dependencies.files.read(approvedPath);
```
[the remark this answers is not in the document]
Comment (777genius) [reply src/features/team-approvals/core/application/use-cases/ReadToolApprovalFilePreview.ts]: Fixed in 24c1d5bf7. The file reader now rejects symlink path components, opens the target with O_NOFOLLOW, binds the initial path stat to the opened descriptor by dev/ino, and revalidates the path identity after reading. Added final symlink, parent symlink, and parent-swap regression tests.

**Comment by 777genius** [comment]
@codex review\n\nPlease review exact head 24c1d5bf7b07e6f7bc27f8d1347e7a6c9ffa9f26. Focus on symlink and TOCTOU resistance, opened-file identity binding, bounded reads, and fail-closed behavior.

File: src/features/team-approvals/main/infrastructure/NodeToolApprovalFileReader.ts [hunk 3]
```diff
@@ -7,20 +8,46 @@ import type { ToolApprovalFileContent } from '@shared/types';
...
+async function lstatSymlinkFreePath(filePath: string): Promise<Stats> {
+  const resolvedPath = path.resolve(filePath);
+  const parsedPath = path.parse(resolvedPath);
+  const segments = resolvedPath.slice(parsedPath.root.length).split(path.sep).filter(Boolean);
+  let currentPath = parsedPath.root;
+  let currentStats = await fs.lstat(currentPath);
+
+  for (const segment of segments) {
+    currentPath = path.join(currentPath, segment);
+    currentStats = await fs.lstat(currentPath);
```
[the remark this answers is not in the document]
Comment (777genius) [reply src/features/team-approvals/main/infrastructure/NodeToolApprovalFileReader.ts]: Fixed in 534b22c89. Linux now walks each component relative to a pinned directory descriptor through /proc/self/fd and opens every component with O_NOFOLLOW. Darwin uses the kernel O_NOFOLLOW_ANY flag for atomic whole-path rejection. Platforms without an equivalent primitive fail closed. The swap regression now proves the unapproved target is never returned.

**Comment by 777genius** [comment]
@codex review\n\nPlease review exact head 534b22c89758f28939801d976db3d80a21c8c78f. Focus on stable directory-handle traversal on Linux, Darwin O_NOFOLLOW_ANY semantics, descriptor lifecycle cleanup, fail-closed unsupported platforms, and parent-swap resistance.

**Comment by 777genius** [comment]
@codex review

**Comment by 777genius** [comment]
@codex review\n\nExact head: f57ddcc006acfbcee51af68cf77824fa38c8985e\n\nPlease review the exact head. Focus on Windows reparse-point and identity checks, Linux procfs traversal, Darwin no-follow behavior, TOCTOU safety, and preview compatibility.

**Comment by 777genius** [comment]
@codex review\n\nPlease review exact head 3a70bef5b after canonical d4cb9a6d synchronization. Focus on the Windows native handle helper, final opened-path binding, FILE_ATTRIBUTE_REPARSE_POINT checks for every component, subprocess containment, and Windows coverage.

**Comment by 777genius** [comment]
@codex review\n\nPlease review exact head d9ada0bb4. Verify ordinal canonical-path binding, symmetric extended-path normalization, no-follow reparse inspection, fail-closed subprocess behavior, and the dedicated Windows CI integration tests.

**Comment by 777genius** [comment]
@codex review

Please review exact head c383b4be3. Recheck embedded C# compilation, ordinal canonical-path binding, symmetric extended-path normalization, no-follow reparse inspection, and the dedicated Windows CI test.

**Comment by 777genius** [comment]
@codex review

Please review exact head 8a4028c0a. The real Windows runner proved helper compilation and all security behaviors; this head only makes the case-mismatch assertion accept PowerShell exception wrapping. Recheck the full exact diff.

**Comment by 777genius** [comment]
@codex review

Please review exact head cffeb78fb73969e8d8c3f6c6abc1d72611f4f854, especially POSIX and Windows path generation binding, missing-parent races, and the hosted-web census migration.

**Comment by 777genius** [comment]
@codex review

Please review exact head 7ad1463324c1cbba5800ecf4dbc23532e82de207. Focus on the PowerShell stdin transport, Windows PowerShell 5.1 compatibility, path encoding and injection resistance, subprocess timeout and output bounds, and the cumulative file-preview security contract.

**Comment by 777genius** [comment]
@codex review

Please review exact head 8eb9f530eb69c4b7b34cd1bc427a2b2ef705310a. Focus on the short PowerShell 5.1 EncodedCommand bootstrap, stdin script execution, EOF and encoding behavior, command-line length, and the cumulative file-preview security contract.

**Comment by 777genius** [comment]
@codex review

**Comment by 777genius** [comment]
@codex review

Please review exact head 16ba49b1449e40fd49c6f71571d1c877a6afa586. Focus on stale or unknown member fail-closed behavior, run and member cwd binding, POSIX and Windows path containment, UTF-16 transport, and backward API compatibility.

**Comment by 777genius** [comment]
@codex review

Please review exact head 6c6478af803131fdfe2ce7962e9d09a7f766a2e9. This replaces commit-topology provenance with a deterministic SHA-256 snapshot over sorted source paths and exact bytes. Please verify squash independence, framing, schema migration, generation, and cumulative approval-preview behavior.

**Comment by 777genius** [comment]
@codex review

Please review exact head 28da7fa79bf4bd09fcb8d9de863fbc5c7dfb8695. This adds a shared core-domain lexical path policy and rejects POSIX parent components, Win32 parent aliases, and every drive-relative qualified path before path resolution. Please verify the cumulative approval-preview security contract and the topology-independent evidence fix.

**Comment by 777genius** [comment]
@codex review

Please review exact head da8c8849b0559fb58551190e94577f84b1c90c25. The lexical policy now rejects all Win32 partially qualified paths before resolution while preserving drive-absolute, complete UNC, extended drive, and extended UNC paths. Please verify remaining Windows qualification edge cases and the cumulative approval-preview security contract.

File: src/features/team-approvals/main/infrastructure/NodeToolApprovalFileReader.ts [hunk 4]
```diff
@@ -1,52 +1,98 @@
...
+  if (segments.length === 0) return fs.open(parsedPath.root, fileFlags);
+
+  const directoryFlags = fileFlags | constants.O_DIRECTORY;
+  let directory = await fs.open(parsedPath.root, directoryFlags);
+  let openedFile: FileHandle | null = null;
+  try {
+    for (const segment of segments.slice(0, -1)) {
+      const nextDirectory = await fs.open(
+        path.join('/proc/self/fd', String(directory.fd), segment),
+        directoryFlags
```
[the remark this answers is not in the document]
Comment (777genius) [reply src/features/team-approvals/main/infrastructure/NodeToolApprovalFileReader.ts]: Fixed. Linux traversal now probes descriptor access and reports unavailable procfs as a fail-closed error instead of a missing file.

File: src/features/team-approvals/main/infrastructure/NodeToolApprovalFileReader.ts [hunk 5]
```diff
@@ -1,52 +1,98 @@
...
+const DARWIN_O_NOFOLLOW_ANY = 0x20000000;
+
+async function openLinuxSymlinkSafePath(filePath: string): Promise<FileHandle> {
+  const parsedPath = path.parse(filePath);
+  const segments = filePath.slice(parsedPath.root.length).split(path.sep).filter(Boolean);
+  const fileFlags = constants.O_RDONLY | constants.O_NOFOLLOW;
+  if (segments.length === 0) return fs.open(parsedPath.root, fileFlags);
+
+  const directoryFlags = fileFlags | constants.O_DIRECTORY;
+  let directory = await fs.open(parsedPath.root, directoryFlags);
```
[the remark this answers is not in the document]
Comment (777genius) [reply src/features/team-approvals/main/infrastructure/NodeToolApprovalFileReader.ts]: Fixed. Linux directory traversal uses O_PATH plus O_DIRECTORY and O_NOFOLLOW, so execute-only parents remain supported.

File: src/features/team-approvals/main/infrastructure/NodeToolApprovalFileReader.ts [hunk 6]
```diff
@@ -1,52 +1,144 @@
...
+    throw error;
+  }
+}
+
+function openSymlinkSafePath(filePath: string): Promise<FileHandle> {
+  if (process.platform === 'darwin') {
+    return fs.open(filePath, constants.O_RDONLY | DARWIN_O_NOFOLLOW_ANY);
+  }
+  if (process.platform === 'linux') return openLinuxSymlinkSafePath(filePath);
+  throw new Error(`Safe approval preview reads are unavailable on ${process.platform}`);
```
[the remark this answers is not in the document]
Comment (777genius) [reply src/features/team-approvals/main/infrastructure/NodeToolApprovalFileReader.ts:107]: Fixed. Windows previews use the native no-follow handle helper, with focused platform tests and hosted Windows CI coverage.

File: src/features/team-approvals/main/infrastructure/NodeToolApprovalFileReader.ts [hunk 7]
```diff
@@ -1,52 +1,176 @@
...
+  try {
+    const parsedPath = path.parse(filePath);
+    const segments = filePath.slice(parsedPath.root.length).split(path.sep).filter(Boolean);
+    let currentPath = parsedPath.root;
+    let pathStats = await fs.lstat(currentPath);
+
+    for (const segment of segments) {
+      currentPath = path.join(currentPath, segment);
+      pathStats = await fs.lstat(currentPath);
+      if (pathStats.isSymbolicLink()) {
```
[the remark this answers is not in the document]
Comment (777genius) [reply src/features/team-approvals/main/infrastructure/NodeToolApprovalFileReader.ts]: Fixed. The native helper opens and validates every Windows component with reparse-point-safe handles.

File: src/features/team-approvals/main/infrastructure/WindowsToolApprovalFileReader.ts [hunk 8]
```diff
@@ -0,0 +1,355 @@
...
+          return new PreviewResult {
+            Exists = true,
+            Content = new byte[0],
+            Error = "Not a file"
+          };
+        }
+
+        string openedPath = NormalizeKernelPath(GetOpenedPath(file));
+        if (!String.Equals(NormalizeComparisonPath(normalizedPath), openedPath, StringComparison.OrdinalIgnoreCase)) {
+          throw new InvalidOperationException("Safe approval preview rejected a redirected Windows path");
```
[the remark this answers is not in the document]
Comment (777genius) [reply src/features/team-approvals/main/infrastructure/WindowsToolApprovalFileReader.ts]: Fixed. Requested and opened Windows paths are normalized into the same namespace, including extended drive and UNC forms.

File: src/features/team-approvals/main/infrastructure/NodeToolApprovalFileReader.test.ts [hunk 9]
```diff
@@ -97,4 +182,186 @@ describe('NodeToolApprovalFileReader', () => {
...
+
+    await expect(windowsReader.read('C:\\linked\\secret.txt')).resolves.toMatchObject({
+      content: '',
+      exists: true,
+      error: expect.stringContaining('reparse-point path'),
+    });
+  });
+
+  it.runIf(process.platform === 'win32')(
+    'reads regular files through the native Windows handle helper',
```
[the remark this answers is not in the document]
Comment (777genius) [reply src/features/team-approvals/main/infrastructure/NodeToolApprovalFileReader.test.ts]: Fixed. The Windows CI smoke compiles and runs the native approval preview suite; the exact head job is green.

File: src/features/team-approvals/main/adapters/input/ipc/registerTeamApprovalsIpc.ts [hunk 10]
```diff
@@ -72,6 +73,37 @@ function validateSettings(settings: unknown): IpcResult<never> | ToolApprovalSet
...
+    return { success: false, error: 'runId must be a non-empty string' };
+  }
+  if (typeof candidate.requestId !== 'string' || candidate.requestId.trim().length === 0) {
+    return { success: false, error: 'requestId must be a non-empty string' };
+  }
+  if (typeof candidate.filePath !== 'string' || candidate.filePath.trim().length === 0) {
+    return { success: false, error: 'filePath must be a non-empty string' };
+  }
+  if (!path.isAbsolute(candidate.filePath)) {
+    return { success: false, error: 'filePath must be an absolute path' };
```
[the remark this answers is not in the document]
Comment (777genius) [reply src/features/team-approvals/main/adapters/input/ipc/registerTeamApprovalsIpc.ts]: Fixed. Relative provider paths are bound to the exact active member or lead project directory while preserving authorization identity.

File: src/features/team-approvals/main/infrastructure/WindowsToolApprovalFileReader.ts [hunk 11]
```diff
@@ -0,0 +1,486 @@
...
+        OPEN_EXISTING,
+        FILE_FLAG_BACKUP_SEMANTICS,
+        IntPtr.Zero
+      );
+    }
+
+    private static SafeFileHandle OpenMissingParentBinding(string filePath) {
+      string parentPath = Path.GetDirectoryName(filePath);
+      SafeFileHandle parent = CreateFileW(
+        parentPath,
```
[the remark this answers is not in the document]
Comment (777genius) [reply src/features/team-approvals/main/infrastructure/WindowsToolApprovalFileReader.ts]: Fixed. Missing Windows parent chains bind through the nearest existing ancestor and remain valid new-file previews.

File: docs/research/hosted-web/phase-0/recovery-events/mutation-census.json [hunk 12]
```diff
@@ -1,10 +1,11 @@
 {
   "schemaVersion": 1,
   "artifactId": "P0.W5.SUPPORTING.MUTATION_CENSUS",
-  "observedAtSha": "668e7d4f2277b727125770960a4d826bf44be6fa",
+  "observedAtSha": "a32f509e6d9bd31ba2135940e336729bf90c3d93",
```
[the remark this answers is not in the document]
Comment (777genius) [reply docs/research/hosted-web/phase-0/recovery-events/mutation-census.json]: Fixed. Census provenance now uses a deterministic digest of every exact source path and byte sequence.

File: src/features/team-approvals/main/infrastructure/WindowsToolApprovalFileReader.ts [hunk 13]
```diff
@@ -0,0 +1,491 @@
...
+  if (!systemRoot || !path.win32.isAbsolute(systemRoot)) {
+    throw new Error('Safe Windows approval preview requires an absolute SystemRoot');
+  }
+  return path.win32.join(systemRoot, 'System32', 'WindowsPowerShell', 'v1.0', 'powershell.exe');
+}
+
+export class PowerShellWindowsPreviewHelper implements WindowsPreviewHelperPort {
+  read(filePath: string): Promise<WindowsPreviewHelperResult> {
+    return new Promise((resolve, reject) => {
+      const encodedPath = Buffer.from(filePath, 'utf8').toString('base64');
```
[the remark this answers is not in the document]
Comment (777genius) [reply src/features/team-approvals/main/infrastructure/WindowsToolApprovalFileReader.ts]: Fixed. Windows path transport is lossless UTF-16LE, including unpaired surrogate code units.

File: docs/research/hosted-web/phase-0/recovery-events/mutation-census.json [hunk 14]
```diff
@@ -1,10 +1,11 @@
 {
   "schemaVersion": 1,
   "artifactId": "P0.W5.SUPPORTING.MUTATION_CENSUS",
-  "observedAtSha": "668e7d4f2277b727125770960a4d826bf44be6fa",
-  "derivation": "TypeScript AST extraction compared bidirectionally with the independently maintained mutation-surface-manifest.json and command descriptors",
+  "observedAtSha": "dab13de404d5f3fbbf8edee15ec361826eec6cb4",
```
[the remark this answers is not in the document]
Comment (777genius) [reply docs/research/hosted-web/phase-0/recovery-events/mutation-census.json]: Fixed. Provenance is now topology-independent sourceSnapshotSha256, so squash and PSC promotion cannot invalidate it.

File: src/main/services/team/provisioning/TeamProvisioningToolApprovalFacade.ts [hunk 15]
```diff
@@ -45,6 +52,71 @@ export interface TeamProvisioningToolApprovalSyncInput extends OpenCodeRuntimeTo
...
+
+function getApprovalPreviewFileTarget(
+  context: PendingToolApprovalFileContext | null
+): PendingToolApprovalFileTarget | null {
+  const authorizationPath = getApprovalPreviewFilePath(context?.approval);
+  if (authorizationPath === null) {
+    return null;
+  }
+  if (path.isAbsolute(authorizationPath)) {
+    return { authorizationPath, readPath: authorizationPath };
```
[the remark this answers is not in the document]
Comment (777genius) [reply src/main/services/team/provisioning/TeamProvisioningToolApprovalFacade.ts]: Fixed. Approval source membership is validated before either absolute or relative targets are accepted.

File: src/main/services/team/provisioning/TeamProvisioningToolApprovalFacade.ts [hunk 16]
```diff
@@ -45,6 +52,71 @@ export interface TeamProvisioningToolApprovalSyncInput extends OpenCodeRuntimeTo
...
+  if (
+    typeof projectDirectory !== 'string' ||
+    projectDirectory.trim().length === 0 ||
+    !path.isAbsolute(projectDirectory)
+  ) {
+    return null;
+  }
+  return {
+    authorizationPath,
+    readPath: path.resolve(projectDirectory, authorizationPath),
```
[the remark this answers is not in the document]
Comment (777genius) [reply src/main/services/team/provisioning/TeamProvisioningToolApprovalFacade.ts:108]: Fixed. Raw parent components and Windows aliases are rejected by the shared domain policy before any path resolution.

File: src/features/team-approvals/core/domain/ToolApprovalPreviewPathPolicy.ts [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+  return true;
+}
+
+export function isToolApprovalPreviewPathLexicallyUnsafe(
+  filePath: string,
+  platform: ToolApprovalPreviewPathPlatform
+): boolean {
+  if (platform === 'win32') {
+    if (/^[a-zA-Z]:(?![\\/])/.test(filePath)) return true;
+    return filePath.split(/[\\/]+/).some(isWindowsParentPathComponent);
```
[the remark this answers is not in the document]
Comment (777genius) [reply src/features/team-approvals/core/domain/ToolApprovalPreviewPathPolicy.ts]: Fixed. Current-drive-rooted and every other partially qualified Windows form are rejected before resolution; exact-head fuzz and hosted review are clean.

**Comment by 777genius** [comment]
@codex review

Exact head: 5a6532c6da124e3de47578dc3535ef5cc5402b7d
Canonical base: b6bfc9a069b5f6067d589760720f672797bf7b61
Please review this exact revision. Focus on POSIX and Windows no-follow path binding, relative-path cwd authorization, reparse and symlink race resistance, fail-closed subprocess behavior, API compatibility, and the cumulative approval-preview security contract.
